src/stdlib_linalg_pinv.fypp

```diff
@@ -0,0 +1,134 @@
+     use stdlib_linalg_constants
+     use stdlib_linalg_blas
+     use stdlib_linalg_lapack
+     use stdlib_linalg_state
+     use ieee_arithmetic, only: ieee_value, ieee_quiet_nan
+     implicit none(type,external)
+     
+     character(*), parameter :: this = 'pseudo-inverse'
+
+     contains
+
+     #:for rk,rt,ri in RC_KINDS_TYPES
+
+     ! Compute the in-place pseudo-inverse of matrix a
+     module subroutine stdlib_linalg_pseudoinvert_${ri}$(a,pinva,rtol,err)
+         !> Input matrix a[m,n]
+         ${rt}$, intent(inout) :: a(:,:)
+         !> Output pseudo-inverse matrix
+         ${rt}$, intent(out) :: pinva(:,:)
+         !> [optional] Relative tolerance for singular value cutoff
+         real(${rk}$), optional, intent(in) :: rtol
+         !> [optional] state return flag. On error if not requested, the code will stop
+         type(linalg_state_type), optional, intent(out) :: err
+
+         ! Local variables
+         real(${rk}$) :: tolerance,cutoff
+         real(${rk}$), allocatable :: s(:)
+         ${rt}$, allocatable :: u(:,:),vt(:,:)
+         type(linalg_state_type) :: err0
+         integer(ilp) :: m,n,k,i,j
+         
+         ! Problem size
+         m = size(a,1,kind=ilp)
+         n = size(a,2,kind=ilp)
+         k = min(m,n)
+         if (m<1 .or. n<1) then
+            err0 = linalg_state_type(this,LINALG_VALUE_ERROR,'invalid matrix size: a=',[m,n])
+            call linalg_error_handling(err0,err)
+            return
+         end if         
+         
+         if (any(shape(pinva,kind=ilp)/=[n,m])) then
+            err0 = linalg_state_type(this,LINALG_VALUE_ERROR,'invalid pinv size:',shape(pinva),'should be',[n,m])
+            call linalg_error_handling(err0,err)
+            return
+         end if                  
+         
+         ! Singular value threshold
+         tolerance = max(m,n)*epsilon(0.0_${rk}$)
+         
+         ! User threshold: fallback to default if <=0
+         if (present(rtol)) then 
+            if (rtol>0.0_${rk}$) tolerance = rtol                
+         end if
+         
+         allocate(s(k),u(m,k),vt(k,n))
+         call svd(a,s,u,vt,overwrite_a=.false.,full_matrices=.false.,err=err0)
+         if (err0%error()) then 
+            err0 = linalg_state_type(this,LINALG_ERROR,'svd failure -',err0%message)
+            call linalg_error_handling(err0,err)
+            return
+         endif
+         
+         !> Discard singular values
+         cutoff = tolerance*maxval(s)
+         s = merge(1/s,0.0_${rk}$,s>cutoff)
+
+         ! Get pseudo-inverse: A_pinv = V * (diag(1/s) * U^H) = V * (U * diag(1/s))^H
+         
+         ! 1) compute (U * diag(1/s)) in-place
+         forall (i=1:m,j=1:k) u(i,j) = s(j)*u(i,j)
+            
+         ! 2) commutate matmul: A_pinv = V^H * (U * diag(1/s))^H = ((U * diag(1/s)) * V^H)^H. 
+         !    This avoids one matrix transpose
+         #:if rt.startswith('complex')
+         pinva = conjg(transpose(matmul(u,vt)))
```

Comment:
These operations could be handled with `gemm` instead for max performance ?

Reply:
[c0f0e10](https://github.com/fortran-lang/stdlib/pull/899/commits/c0f0e10e31a7c0a5933a0e2e0c9d06dad5b22ffc)